File: core/dashboard/entity_routing.py
"""Resolve HA-style entity card preset to a concrete dashboard renderer."""

from __future__ import annotations

from typing import Any

from core.ui_catalog import _CARD_RENDERERS

_TOGGLE_DOMAINS = frozenset({"switch", "input_boolean"})

_DEDICATED_DOMAIN_RENDERERS: dict[str, str] = {
    "number": "number",
    "select": "select",
    "sensor": "sensor",
    "binary_sensor": "sensor",
    "light": "light",
    "climate": "climate",
    "lock": "lock",
    "vacuum": "vacuum",
    "weather": "weather",
    "scene": "scene",
    "button": "button",
    "script": "button",
}

_TILE_DOMAINS = frozenset({"cover", "fan", "media_player"})


def _entity_domain(widget: dict[str, Any] | None) -> str:
    if not isinstance(widget, dict):
        return ""
    domain = str(widget.get("domain") or "").strip().lower()
    if domain:
        return domain
    entity_id = str(widget.get("entity_id") or "").strip()
    if "." in entity_id:
        return entity_id.split(".", 1)[0].lower()
    return ""


def resolve_entity_effective_renderer(widget: dict[str, Any] | None) -> dict[str, Any]:
    if not isinstance(widget, dict):
        return {"renderer": "info", "switch_style": False}

    stored = str(widget.get("renderer") or "").strip().lower()
    if stored and stored not in {"entity", ""} and stored in _CARD_RENDERERS:
        return {
            "renderer": stored,
            "switch_style": bool(widget.get("switch_style")),
        }

    domain = _entity_domain(widget)
    switch_style = bool(widget.get("switch_style"))

    dedicated = _DEDICATED_DOMAIN_RENDERERS.get(domain)
    if dedicated:
        return {"renderer": dedicated, "switch_style": False}

    if domain in _TOGGLE_DOMAINS or switch_style:
        return {"renderer": "switch", "switch_style": True}
    if domain in _TILE_DOMAINS:
        return {"renderer": "tile", "switch_style": False}
    return {"renderer": "info", "switch_style": False}

File: static/hyveview/core/entity_routing.ts
/**
 * Resolve universal "entity" card preset to a concrete Hyveview renderer.
 */

import type { HyveviewWidget } from '../types/widget.js';

const TOGGLE_DOMAINS = new Set(['switch', 'input_boolean']);

const DEDICATED_DOMAIN_RENDERERS: Record<string, string> = {
    number: 'number',
    select: 'select',
    sensor: 'sensor',
    binary_sensor: 'sensor',
    light: 'light',
    climate: 'climate',
    lock: 'lock',
    vacuum: 'vacuum',
    lawn_mower: 'lawn_mower',
    weather: 'weather',
    scene: 'scene',
    button: 'button',
    script: 'button',
};

const TILE_DOMAINS = new Set([
    'cover', 'fan', 'media_player',
]);

const RESOLVED_RENDERERS = new Set([
    'number', 'select', 'sensor', 'scene', 'button', 'switch', 'tile', 'info',
    'light', 'climate', 'gauge', 'lock', 'vacuum', 'lawn_mower', 'weather', 'weather_rich',
    'camera', 'picture', 'fusion_solar', 'label',
]);

export function entityDomainFromWidget(widget: HyveviewWidget | null | undefined): string {
    const domain = String(widget?.domain || '').trim().toLowerCase();
    if (domain) return domain;
    const entityId = String(widget?.entity_id || '').trim();
    const dot = entityId.indexOf('.');
    return dot > 0 ? entityId.slice(0, dot).toLowerCase() : '';
}

export function resolveEntityEffectiveType(
    widget: HyveviewWidget | null | undefined,
): { effectiveType: string; switchStyle: boolean } {
    const storedRenderer = String(widget?.renderer || '').trim().toLowerCase();
    if (storedRenderer && storedRenderer !== 'entity' && RESOLVED_RENDERERS.has(storedRenderer)) {
        return {
            effectiveType: storedRenderer,
            switchStyle: Boolean(widget?.switch_style),
        };
    }

    const domain = entityDomainFromWidget(widget);
    const switchStyle = Boolean(widget?.switch_style);

    const dedicated = DEDICATED_DOMAIN_RENDERERS[domain];
    if (dedicated) return { effectiveType: dedicated, switchStyle: false };

    if (TOGGLE_DOMAINS.has(domain)) {
        return { effectiveType: 'switch', switchStyle: false };
    }
    if (switchStyle) {
        return { effectiveType: 'switch', switchStyle: true };
    }
    if (TILE_DOMAINS.has(domain)) {
        return { effectiveType: 'tile', switchStyle: false };
    }
    return { effectiveType: 'info', switchStyle: false };
}
